File: game.py
import websockets #used for communication with browser
import asyncio
import aiohttp #used for communication with score microservice
import json
import math
import time
#adding async before a function definition adds the function to the event loop

#Game area (defined in game.html)
width = 800
height = 500
thrust_power = 0.3
ship_size = 50
score_to_win = 10
max_players = 2

spawn_points = [
    (100, 250),
    (700, 250)
]

#Players
players = {}

connected_players = {}
game_state = {
    'players': {},
}

game_start_time = None

# This is called every time a new player connects
async def player_connection(websocket):
    if len(game_state['players']) >= max_players:
        await websocket.send(json.dumps({
            'type': 'error',
            'message': 'Game is full. Try again later.'
        }))
        await websocket.close()
        print('Rejected a player connection')
        return
    
    init_data = json.loads(await websocket.recv())
    screen_name = init_data.get('screen_name','Unknown')
    player_id = init_data.get('player_id')
    print(screen_name + ' connected')

    #Add player to connected players
    connected_players[player_id] = websocket

    setup_message = {
        'type': 'setup',
        'screen_name': screen_name,
        'ship_size': ship_size,
        'players' : 1, # temporary. will need to send message with all player ids
    }
    
    # Send setup message to browser
    await websocket.send(json.dumps(setup_message))

    # Initialize player score
    asyncio.create_task(init_player_score(player_id))

    # Add player to the game
    spawn_index = len(game_state['players'])
    x, y = spawn_points[spawn_index]
    game_state['players'][player_id] = {
        'screen_name': screen_name,
        'spawn_index': spawn_index,
        'x':x, 
        'y':y, 
        'vx':0, # starting velocity
        'vy':0, # starting velocity
        'rotation':0, #starting rotation
        'laser': {
            'active': False,
            'cooldown': False,
            'duration': 3,
            'hitTarget': False,
            'from': {
                'x': 0,
                'y': 0
            },
            'to': {
                'x': 0,
                'y': 0
            },
        },
        'keys': {
            'rotateLeft': False,
            'rotateRight': False,
            'thrust': False,
            'fireLaser': False
        }
    }

    game_state_message = {
        'type':'game_update',
        'players': game_state['players'],
    }

    # Send starting game state to new player
    await websocket.send(json.dumps(game_state_message))

    # If second player just joined, broadcast game_start to everyone
    if len(game_state['players']) == 2:
        global game_start_time
        game_start_time = time.time()

        # Reset players to spawn points
        for pid, player in game_state['players'].items():
            x, y = spawn_points[player['spawn_index']]
            player['x'] = x
            player['y'] = y
            player['vx'] = 0
            player['vy'] = 0
            player['rotation'] = 0

        await broadcast_message({
            'type': 'game_start',
            'players': game_state['players']
        })

    # Listen for messages from this player
    try:
        async for message in websocket:
            await handle_player_message(player_id, message)

    except websockets.exceptions.ConnectionClosed:
        print(f'Player {player_id} disconnected')

    finally:
        # Handle Disconnection
        screen_name = game_state['players'][player_id]['screen_name']
        del game_state['players'][player_id]
        del connected_players[player_id]
        await broadcast_message({
            'type': 'game_update',
            'players': game_state['players']
            })
        
        # Send message to remaining players
        message = json.dumps({
            'type': 'player_disconnected',
            'player_name': screen_name
        })
        for websocket in connected_players.values():
            await websocket.send(message)

        await reset_game()

# Process messages from browser
async def handle_player_message(player_id, message):
    data = json.loads(message)
    if data['type'] == 'input':
        keys = data['keys']
        player = game_state['players'][player_id]
        player['keys'] = keys

async def broadcast_message(message):
    for websocket in connected_players.values():
        try:
            await websocket.send(json.dumps(message))
        except:
            break

def get_laser_endpoint(from_x, from_y, angle_rad):
    to_x = from_x
    to_y = from_y
    #Keep extending line until it crosses boundary
    while to_x >= 0 and to_x <= width and to_y >= 0 and to_y <= height:
        to_x += math.cos(angle_rad) * 5
        to_y += math.sin(angle_rad) * 5
    return (to_x, to_y)

def laser_hit(laser_from, laser_to, ship_x, ship_y):
    # Use algorithm for checking if a line intersects a circle
    dx = laser_to['x'] - laser_from['x']
    dy = laser_to['y'] - laser_from['y']
    fx = laser_from['x'] - ship_x
    fy = laser_from['y'] - ship_y

    a = dx**2 + dy**2
    b = 2 * (fx * dx + fy * dy)
    c = fx**2 + fy**2 - ship_size**2

    d = b**2 - 4 * a * c

    # No intersection
    if d < 0:
        return False
    
    d = math.sqrt(d)
    t1 = (-b - d) / (2*a)
    t2 = (-b + d) / (2*a)

    # Intersection
    if (t1 >= 0 and t1 <= 1) or (t2 >= 0 and t2 <= 1):
        return True
    # No intersection
    else:
        return False
    
async def init_player_score(player_id):
    async with aiohttp.ClientSession() as session:
        try:
            await session.post(
                'http://localhost:8767/score/init',
                json={'player_id': player_id}
            )
        except:
            print('Score initialization failed')
    
async def send_score_update(player_id):
    async with aiohttp.ClientSession() as session:
        try:
            response = await session.post(
                'http://localhost:8767/score/hit',
                json={'player_id': player_id}
            )
            data = await response.json()
            if data['score'] >= score_to_win:
                await broadcast_game_over(player_id)
        except:
            print('Score update failed')

async def broadcast_game_over(winner_id):
    winner = game_state['players'][winner_id]
    message = json.dumps({
        'type': 'game_over',
        'winner_name': winner['screen_name']
    })
    for websocket in connected_players.values():
        await websocket.send(message)

    await reset_game()

async def reset_game():
    global current_id
    current_id = 1
    game_state['players'].clear()
    connected_players.clear()
    
    # Reset scores
    async with aiohttp.ClientSession() as session:
        try:
            await session.post('http://localhost:8767/score/reset')
            await session.post('http://localhost:8766/login/reset')
        except:
            print('Reset failed')


async def game_loop():
    while True:
        countdown_active = game_start_time and (time.time() - game_start_time < 4)
        if not countdown_active:
            for player_id, player in game_state['players'].items():
                keys = player['keys']

                if keys.get('rotateLeft'):
                    player['rotation'] -= 5
                if keys.get('rotateRight'):
                    player['rotation'] += 5
                if keys.get('thrust'):
                    angle_rad = math.radians(player['rotation'] - 90)
                    player['vx'] += math.cos(angle_rad) * thrust_power
                    player['vy'] += math.sin(angle_rad) * thrust_power
                if keys.get('fireLaser') and not player['laser']['cooldown']:
                        player['laser']['active'] = True
                        player['laser']['hitTarget'] = False
                        player['laser']['duration'] = 3
                        player['laser']['cooldown'] = 10

                # Update position
                player['x'] += player['vx']
                player['y'] += player['vy']

                # Apply friction
                player['vx'] *= 0.99
                player['vy'] *= 0.99

                # Keep within bounds
                if player['x'] - (ship_size/2) < 0:
                    player['x'] = (ship_size/2)
                    player['vx'] = 0
                elif player['x'] + (ship_size/2) > width:
                    player['x'] = width - (ship_size/2)
                    player['vx'] = 0

                if player['y'] - (ship_size/2) < 0:
                    player['y'] = (ship_size/2)
                    player['vy'] = 0
                elif player['y'] + (ship_size/2) > height:
                    player['y'] = height - (ship_size/2)
                    player['vy'] = 0

                if player['laser']['active']:
                    # Update laser position
                    angle_rad = math.radians(player['rotation'] - 90)
                    from_x = player['x'] + (ship_size/2) * math.cos(angle_rad)
                    from_y = player['y'] + (ship_size/2) * math.sin(angle_rad)
                    (to_x, to_y) = get_laser_endpoint(from_x, from_y, angle_rad)
                    player['laser']['from']['x'] = from_x
                    player['laser']['from']['y'] = from_y
                    player['laser']['to']['x'] = to_x
                    player['laser']['to']['y'] = to_y

                    # Update laser duration
                    player['laser']['duration'] -= 1
                    if player['laser']['duration'] <= 0:
                        player['laser']['active'] = False

                    # Check for laser hit
                    if not player['laser']['hitTarget']: # A laser can only hit a target once
                        for other_player in game_state['players'].values():
                            if player == other_player:
                                continue # Don't check if player hits their own laser
                            else:
                                if laser_hit(player['laser']['from'], player['laser']['to'], other_player['x'], other_player['y']):
                                    print('Hit detected')
                                    player['laser']['hitTarget'] = True
                                    asyncio.create_task(send_score_update(player_id))

                # Update laser cooldown timer
                if player['laser']['cooldown']:
                    player['laser']['cooldown'] -= 1
                    if player['laser']['cooldown'] <= 0:
                        player['laser']['cooldown'] = False
                

        # Send current game state to all players
        await broadcast_message({
            'type': 'game_update',
            'players': game_state['players']
            })
        
        await asyncio.sleep(1 / 30)  # 30 FPS


async def main():
    server = await websockets.serve(player_connection, 'localhost', 8765) #starts the server. Player connection is the function that is run whenever a new connection is received
    print('Game server is running')
    await asyncio.gather(
        server.wait_closed(),
        game_loop()
    )


if __name__ == "__main__":
    asyncio.run(main())

File: game.js
// Game variables
let ship_size = null;
let myPlayerID = null;
let myScreenName = null;
let gameStartTime = null;
let gameState = {
    players: {}
};

let keys = {
    rotateLeft: false,
    rotateRight: false,
    thrust: false,
    fireLaser: false
}

// Spaceship Images
let mySpaceshipImage = null;
let otherSpaceshipImage = null;

// Canvas Setup
var canvas = document.getElementById('canvas');
var g = canvas.getContext('2d'); 

// Chat variables
let chatSocket = null;
let chatMessages = []; // store chat history

function loadImage(src) {
    return new Promise((resolve) => {
        const img = new Image();
        img.onload = () => resolve(img);
        img.src = src;
    });
}

function connect() {
    websocket = new WebSocket('ws://localhost:8765'); //address of server

    websocket.onopen = function() {
        console.log('Connected to game server');
        websocket.send(JSON.stringify({
            type: 'init',
            screen_name: sessionStorage.getItem('screen_name'),
            player_id: sessionStorage.getItem('player_id')
        }))
    }

    websocket.onmessage = function(event) {
        const message = JSON.parse(event.data);
        handleMessage(message);
    }
}

function handleMessage(message) {
    switch (message.type) {
        case 'setup':
            myPlayerID = sessionStorage.getItem('player_id')
            myScreenName = message.screen_name;
            ship_size =  message.ship_size;
            gameState.players = message.players;
            console.log('Received setup message');
            startInputLoop();
            break;

        case 'game_start':
            gameStartTime = Date.now();
            gameState.players = message.players;
            renderGame();
            break;
        
        case 'game_update':
            gameState.players = message.players;
            renderGame();
            break;

        case 'player_disconnected':
            alert(`${message.player_name} disconnected. Returning to menu.`);
            window.location.href = 'index.html';
            break;

        case 'game_over':
            alert(`${message.winner_name} wins!`);
            window.location.href = 'index.html';
            break;
    }
}

function startInputLoop() {
    // Send input to server at 30 FPS
    setInterval(() => {
        if (websocket && websocket.readyState === WebSocket.OPEN) {
            websocket.send(JSON.stringify({
                type: 'input',
                keys: keys,
                timestamp: Date.now()
            }));
        }
    }, 1000/30); // 30 FPS
}

function renderGame() {
    g.clearRect(0,0,canvas.clientWidth,canvas.clientHeight)

    if (Object.keys(gameState.players).length < 2) {
        g.fillStyle = 'white';
        g.font = '30px "Lucida Console", Monaco, monospace';
        g.textAlign = 'center';
        g.fillText('Waiting for Player 2...', canvas.width / 2, canvas.height / 2);
    }

    if (gameStartTime && (Date.now() - gameStartTime < 3000)) {
        const countdown = Math.ceil(3 - (Date.now() - gameStartTime) / 1000)
        g.fillStyle = 'white';
        g.font = '60px "Lucida Console", Monaco, monospace';
        g.textAlign = 'center';
        g.fillText(countdown, canvas.width / 2, canvas.height / 2);
    }

    // Draw all players and lasers
    for (let playerID in gameState.players) {
        const player = gameState.players[playerID];
        drawPlayer(player, playerID == myPlayerID);
        drawLaser(player);
    }
}

function drawPlayer(player, isMe) {
    g.save();
    g.translate(player.x, player.y); //player position
    g.rotate(player.rotation * Math.PI/180); //convert to radians
    const img = isMe ? mySpaceshipImage : otherSpaceshipImage;
    g.drawImage(
        img,
        -ship_size/2, //-25, //negative half width
        -ship_size/2, //-25, //negative half height
        ship_size, //50, //spaceship width
        ship_size //50 //spaceship height
        );
    g.restore();
}

function drawLaser(player) {
    if (player.laser.active) {
        g.strokeStyle = 'red';
        g.lineWidth = 2;
        g.beginPath()
        g.moveTo(player.laser.from.x, player.laser.from.y);
        g.lineTo(player.laser.to.x, player.laser.to.y);
        g.stroke();
    }
}

async function fetchScores() {
    try {
        const response = await fetch('http://localhost:8767/score');
        if (!response.ok) throw new Error('Failed to fetch scores');
        const scores = await response.json();
        updateScoreboard(scores);
    }
    catch (e) {
        console.error('Error fetching scores: ', e);
    }
}

function updateScoreboard(scores) {
    const scoreboard = document.getElementById('scoreboard');
    let html = ''

    let playerScoresHtml = [];
    for (let playerID in scores) {
        const player = gameState.players[playerID];
        const name = player.screen_name;
        // Wrap each player's score in a span with a class
        playerScoresHtml.push(`<span class="player-score">${name}: ${scores[playerID]}</span>`);
    }
    // Join them without additional separators, as flexbox will handle spacing
    html += playerScoresHtml.join('');
    scoreboard.innerHTML = html;
}

// Keyboard input handling
document.addEventListener('keydown', function(event) {
    switch(event.key) {
        case 'f':
            console.log('keyboard input received')
            keys.thrust = true;
            break;
        case 'j':
            console.log('keyboard input received')
            keys.rotateLeft = true;
            break;
        case 'l':
            console.log('keyboard input received')
            keys.rotateRight = true;
            break;
        case ' ':
            console.log('keyboard input received')
            keys.fireLaser = true;
            break;
    }
});

document.addEventListener('keyup', function(event) {
    switch(event.key) {
        case 'f':
            keys.thrust = false;
            break;
        case 'j':
            keys.rotateLeft = false;
            break;
        case 'l':
            keys.rotateRight = false;
            break;
        case ' ':
            keys.fireLaser = false;
            break;
    }
});

function connectChat() {
    chatSocket = new WebSocket('ws://localhost:8768');

    chatSocket.onopen = () => {
        console.log('Connected to chat server');
    };

    chatSocket.onmessage = (event) => {
        console.log('chat received')
        const message = JSON.parse(event.data);

        switch(message.type) {
            case 'chat_history':
                chatMessages = message.messages;
                renderChat();
                break;

            case 'chat_message':
                chatMessages.push({
                    player_id: message.player_id,
                    message: message.message,
                    timestamp: message.timestamp
                });
                renderChat();
                break;
        }
    };

    chatSocket.onclose = () => {
        console.log('Chat server connection closed');
    };
}

function sendChatMessage(text) {
    if (chatSocket && chatSocket.readyState === WebSocket.OPEN) {
        console.log('Sending chat')
        chatSocket.send(JSON.stringify({
            type: 'chat_message',
            player_id: myPlayerID,
            message: text
        }));
    }
}

function renderChat() {
    const chatbox = document.getElementById('chatbox');
    chatbox.innerHTML = chatMessages.map(msg => {
        const player = gameState.players[msg.player_id];
        const name = player.screen_name;
        return `<div><strong>${name}:</strong> ${msg.message}</div>`;
    }).join('');
}


// This is the entry point.  It runs when the webpage finishes loading
window.addEventListener('load', async function() {
    fetchScores(); // Show scores before connecting to game server

    // Force it to load images before connecting to game server
    [mySpaceshipImage, otherSpaceshipImage] = await Promise.all([
        loadImage('Images/my_spaceship.png'),
        loadImage('Images/other_spaceship.png')
    ]);
    connect(); //Connect to game server
    this.setInterval(fetchScores, 1000); //Update scores once per second

    //Connect chat buttons
    const chatInput = document.getElementById('chat-input');
    const sendButton = document.getElementById('chat-send');

    sendButton.onclick = () => {
        if (chatInput.value.trim()) {
            sendChatMessage(chatInput.value.trim());
            chatInput.value = '';
        }
    };

    chatInput.addEventListener('keypress', (e) => {
        if (e.key === 'Enter') {
            sendButton.click();
        }
    });

    connectChat(); //Connect to chat server


});